### PR TITLE
Tunable dense_layers in GraphConvModel

### PR DESCRIPTION
**Pull request by cpfpengfei**
Instead of a single dense layer in GraphConvModel, this PR proposes a simple edit to have multiple dense layers in a list as an argument. This is my first PR hoping to contribute to DeepChem, and open to further edits / fine-tuning, thank you!

### CODE REVIEW

File: deepchem/models/graph_models.py [hunk 1]
```diff
@@ -715,7 +715,7 @@ class _GraphConvKerasModel(tf.keras.Model):
   def __init__(self,
                n_tasks,
                graph_conv_layers,
-               dense_layer_size=128,
+               dense_layers, # changed dense_layer_size to dense_layers as a list instead
```
Comment (rbharath) [inline deepchem/models/graph_models.py:718]: Could you add type hints for this class? Check out our coding conventions page which provides an overview of our recent transition to type hints: https://deepchem.readthedocs.io/en/latest/coding.html

File: deepchem/models/graph_models.py [hunk 2]
```diff
@@ -715,7 +715,7 @@ class _GraphConvKerasModel(tf.keras.Model):
   def __init__(self,
```
Comment (rbharath) [inline deepchem/models/graph_models.py:715]: Could you add a unit test for `GraphConvModel` that makes use of the dense layers?

File: deepchem/models/graph_models.py [hunk 3]
```diff
@@ -715,7 +715,7 @@ class _GraphConvKerasModel(tf.keras.Model):
...
                n_tasks,
                graph_conv_layers,
-               dense_layer_size=128,
+               dense_layers, # changed dense_layer_size to dense_layers as a list instead
                dropout=0.0,
```
Comment (rbharath) [inline deepchem/models/graph_models.py:719]: Could you add options for weight/bias initialization, batch normalization, and weight decay for fully connected layers as in `WeaveModel`? (https://github.com/deepchem/deepchem/blob/master/deepchem/models/graph_models.py#L89)

File: deepchem/models/graph_models.py [hunk 4]
```diff
@@ -754,15 +754,18 @@ def __init__(self,
         layers.GraphConv(layer_size, activation_fn=tf.nn.relu)
         for layer_size in graph_conv_layers
     ]
+    self.dense_layers = [
+        Dense(layer_size, activation = tf.nn.relu)
```
Comment (rbharath) [inline deepchem/models/graph_models.py:758]: Can you add a argument for `activation_fn` so the user can set their choice of activation function?

File: deepchem/models/graph_models.py (hunk 3 above)
Comment (rbharath) [reply deepchem/models/graph_models.py:719]: Ah we already have batch-norms and dropouts so we're good on that front

File: deepchem/models/graph_models.py [hunk 5]
```diff
@@ -794,12 +797,18 @@ def call(self, inputs, training=False):
...
-    if training and self.dropouts[-1] is not None:
-      dense = self.dropouts[-1](dense, training=training)
-    neural_fingerprint = self.graph_gather([dense, degree_slice, membership] +
+
+    startnum = len(self.graph_convs)
+    endnum = len(self.dense_layers) + len(self.graph_convs)
+    for i in range(startnum, endnum): # cont index for batchnorms and dropouts in dense layers
+        dense = self.dense_layers[i-startnum](in_layer)
+        if self.batch_norms[i] is not None:
+            dense = self.batch_norms[i](dense, training = training)
```
Comment (rbharath) [inline deepchem/models/graph_models.py:806]: Can you add `batch_normalize_kwargs` as in `WeaveModel` so users can configure their batch norms? https://github.com/deepchem/deepchem/blob/master/deepchem/models/graph_models.py#L96

File: deepchem/models/graph_models.py [hunk 6]
```diff
@@ -740,8 +740,8 @@ def __init__(self,
     self.uncertainty = uncertainty
```
Comment (rbharath) [inline deepchem/models/graph_models.py:741]: Could you add a usage example to the docstring?

**Comment by cpfpengfei** [comment]
@rbharath Hi Bharath, thanks for all your clarifications and your code review! I will find time to work on this added tunable dense layers feature for ```GraphConvModel``` soon. However, before so, I do actually have some more specific questions regarding ```GraphConvModel``` to further understand what is going on, hope you can help me with them during your free time!

1. In ```graph_models.py``` for ```GraphConvModel``` model class, the description states ```graph_conv_layers: list of int Width of channels for the Graph Convolution Layers```. Usually, the ```graph_conv_layer``` has a trainable parameter in the form of a weight matrix. So assuming it is ```H*W``` since the first dimension is decided by the input dimension, does this list of ```int``` determines the second dimension so the feature vector dimension of each node can be changed? Or is it that these two dimensions are fixed, and this argument actually means channel num? So we may have different number of graph convolution filters? If it is referring to the W dimension, does it mean that we don’t have a choice to have multiple number of graph convolutional filters, and there is only one for each layer?

2. Furthermore, for the dense layer in the graph model, the description states that it is ```Width of channels for Atom Level Dense Layer before GraphPool```. Similarly, what does the width actually mean? Also, I am confused about the position of the dense layer, since according to the code, it seems like the dense layer is before the creation of the ```neural_fingerprint`` but then there is another dense layer after the fingerprint has been formed to do the downstream regression task. As I thought that the structure is as such: ```graph_conv + graph_pool + graph_conv + graph _pool + graph_gather = graph level representation```. Then from this representation, we use another dense NN to do further tasks.

3. The ```self.regression_dense = Dense(n_tasks) in class _GraphConvKerasModel(tf.keras.Model)``` is confusing as well. For example in the case of 1 prediction task, so ```n_tasks = 1```, does it means that the dense layer only has one node?
```output = self.regression_dense(neural_fingerprint)```, then for this, it means that after graph gathering, we only use one node to output the fingerprint results? Basically it is just summing everything out with a trainable weight. So the whole prediction is only based on the fingerprint instead of the output representation and then we use a dense layer for prediction?

4. Lastly, for the ```dense_layer```, although the description says that it is before ```GraphPool```, in the code implementation it seems to work after the ```GraphPool``` but before the ```GraphGather``` segment. Why do we need the dense layer before the graph gather layer? 

Thanks a lot for your time!

**Comment by rbharath** [comment]
> @rbharath Hi Bharath, thanks for all your clarifications and your code review! I will find time to work on this added tunable dense layers feature for `GraphConvModel` soon. However, before so, I do actually have some more specific questions regarding `GraphConvModel` to further understand what is going on, hope you can help me with them during your free time!
> 
> 1. In `graph_models.py` for `GraphConvModel` model class, the description states `graph_conv_layers: list of int Width of channels for the Graph Convolution Layers`. Usually, the `graph_conv_layer` has a trainable parameter in the form of a weight matrix. So assuming it is `H*W` since the first dimension is decided by the input dimension, does this list of `int` determines the second dimension so the feature vector dimension of each node can be changed? Or is it that these two dimensions are fixed, and this argument actually means channel num? So we may have different number of graph convolution filters? If it is referring to the W dimension, does it mean that we don’t have a choice to have multiple number of graph convolutional filters, and there is only one for each layer?

So the `graph_conv_layer` parameters determine the size of the per-atom output vector at each layer. So if we have `[64, 128]`, after the first GraphConv layer, each atom will have a 64 dimensional feature vector, and after the 2nd, a 128 dimensional feature vector. The weight matrices will be sized accordingly to ensure outputs of this size 

> 2. Furthermore, for the dense layer in the graph model, the description states that it is `Width of channels for Atom Level Dense Layer before GraphPool`. Similarly, what does the width actually mean? Also, I am confused about the position of the dense layer, since according to the code, it seems like the dense layer is before the creation of the `neural_fingerprint` but then there is another dense layer after the fingerprint has been formed to do the downstream regression task. As I thought that the structure is as such: `graph_conv + graph_pool + graph_conv + graph _pool + graph_gather = graph level representation`. Then from this representation, we use another dense NN to do further tasks.

This is a good question, and I'm not entirely sure why we have this dense layer. I'd recommend leaving it in for now. I'll think about this more and get back to you.

> 3. The `self.regression_dense = Dense(n_tasks) in class _GraphConvKerasModel(tf.keras.Model)` is confusing as well. For example in the case of 1 prediction task, so `n_tasks = 1`, does it means that the dense layer only has one node?
>    `output = self.regression_dense(neural_fingerprint)`, then for this, it means that after graph gathering, we only use one node to output the fingerprint results? Basically it is just summing everything out with a trainable weight. So the whole prediction is only based on the fingerprint instead of the output representation and then we use a dense layer for prediction?

The final dense layer is used to reshape the output of the graph gather to the shape required for the prediction essentially.

> 4. Lastly, for the `dense_layer`, although the description says that it is before `GraphPool`, in the code implementation it seems to work after the `GraphPool` but before the `GraphGather` segment. Why do we need the dense layer before the graph gather layer?
> 
> Thanks a lot for your time!

Hmm, good question. I'll need to go back and read the source Duvenaud paper to see if we match it exactly. For now, I'd recommend leaving this as is and just adding the fully connected layers after graph gather and before the final dense layer to the outputs. It should be possible to choose to have no fully connected layers and retain the original model that we have currently if users so desire.

**Comment by cpfpengfei** [comment]
@rbharath Hi Bharath, thanks for your replies here and on the Gitter community! Carried over from Gitter, here are our questions on prediction dense layers and involvement of neural_fingerprint on uncertainty calculations, particularly on ```class _GraphConvKerasModel(tf.keras.Model):```

```python
    else:
      output = self.regression_dense(neural_fingerprint)
      output = self.trim([output, n_samples])
      if self.uncertainty:
        log_var = self.uncertainty_dense(neural_fingerprint)
        log_var = self.uncertainty_trim([log_var, n_samples])
        var = self.uncertainty_activation(log_var)
        outputs = [output, var, output, log_var, neural_fingerprint]
      else:
        outputs = [output, neural_fingerprint]
```

Why is ```neural_fingerprint``` an input for uncertainty_dense() instead of the final ```output``` being the input this way: self.uncertainty_dense(output)?

If I were to add more dense layers after neural_fingeprint (instead of a single node right now), the uncertainty output will have to be reconstructed differently instead of reading from neural_fingerprint?

Regarding your reply that the "neural_fingerprint should be the output of the final dense layer (if we have dense layers)", why is this so? Isn't neural_fingerprint the fingerprint of molecular structures after the graph convolution and pooling layers and thus should be passed to dense layers thereafter? Apologies if I have a misconception here!

@jiali1025 has been working closely with me and would like to ask some questions to understand better on the implementation as well. Thank you very much!

**Comment by jiali1025** [comment]
@rbharath Hi Bharath，I am jiali. I would like to add some arguments to pengfei's question for asking you. I have gone through several papers of spatial-based graph convolutional network and in my understanding, the graph conv layers are aiming to transmit information between nodes to get better node features that include their neighbor's information. 

After the graph conv layer, we get a new feature representation of each node. However, normally the molecular prediction is a graph-level task, so we have the graph-gathering layer which normally is a sum or average operation. Then we can get the neural-fingerprint (new feature vector) for each molecular graph. Which we tested with your code, the neural fingerprint is a 205 * 1 dimension vector for our case. 

So for the uncertainty calculation here if use neural fingerprint is a bit strange. Since the final task is a property prediction for a single value. So the uncertainty in model's prediction should be of that single value, the uncertainty in a vector seems to be strange. Is this because your implementation of the uncertainty calculation class, since some of the codes I don't perfectly understand, so is not able to figure the problem out now. 


Many thanks!

Jiali Li

**Comment by rbharath** [comment]
Sorry for the late reply here! 

> ```python
>     else:
>       output = self.regression_dense(neural_fingerprint)
>       output = self.trim([output, n_samples])
>       if self.uncertainty:
>         log_var = self.uncertainty_dense(neural_fingerprint)
>         log_var = self.uncertainty_trim([log_var, n_samples])
>         var = self.uncertainty_activation(log_var)
>         outputs = [output, var, output, log_var, neural_fingerprint]
>       else:
>         outputs = [output, neural_fingerprint]
> ```
> 
> Why is `neural_fingerprint` an input for uncertainty_dense() instead of the final `output` being the input this way: self.uncertainty_dense(output)?

Hmm, I'm not sure off hand. @peastman would you happen to know?

I'll also try to take a closer look at this later

> If I were to add more dense layers after neural_fingeprint (instead of a single node right now), the uncertainty output will have to be reconstructed differently instead of reading from neural_fingerprint?

Hmm, I think you're right. I think the uncertainty output should be constructed from the final dense layer output basically .
> 
> Regarding your reply that the "neural_fingerprint should be the output of the final dense layer (if we have dense layers)", why is this so? Isn't neural_fingerprint the fingerprint of molecular structures after the graph convolution and pooling layers and thus should be passed to dense layers thereafter? Apologies if I have a misconception here!

It's a good question! I think either representation has a reasonable cause to be called the neural fingerprint. I'd tentatively lean towards being the output of the dense layers but can see if being the output from the graph convolutions as well

**Comment by rbharath** [comment]
> @rbharath Hi Bharath，I am jiali. I would like to add some arguments to pengfei's question for asking you. I have gone through several papers of spatial-based graph convolutional network and in my understanding, the graph conv layers are aiming to transmit information between nodes to get better node features that include their neighbor's information.
> 
> After the graph conv layer, we get a new feature representation of each node. However, normally the molecular prediction is a graph-level task, so we have the graph-gathering layer which normally is a sum or average operation. Then we can get the neural-fingerprint (new feature vector) for each molecular graph. Which we tested with your code, the neural fingerprint is a 205 * 1 dimension vector for our case.
> 
> So for the uncertainty calculation here if use neural fingerprint is a bit strange. Since the final task is a property prediction for a single value. So the uncertainty in model's prediction should be of that single value, the uncertainty in a vector seems to be strange. Is this because your implementation of the uncertainty calculation class, since some of the codes I don't perfectly understand, so is not able to figure the problem out now.

This is a good question @jiali1025! My honest answer is that I don't know right now whether it makes more sense to have the uncertainty be computed at the output. There may be a technical reason we have it as-is right now that I'm not remembering off-hand. My two cents is it's probably better to follow the current code structure unless we can articulate a clear reason we should change the uncertainty handling structure

**Comment by peastman** [comment]
> Why is neural_fingerprint an input for uncertainty_dense() instead of the final output being the input this way: self.uncertainty_dense(output)?

It's making two independent predictions: the value of the output, and also the uncertainty in that value.  The uncertainty isn't just a function of the output value.  Two different inputs might lead to exactly the same output but with different uncertainties.  See https://arxiv.org/abs/1703.04977 for a description of how this works.
